go/mysql/mysql56_gtid_set.go

```diff
@@ -435,6 +435,128 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
 	return buf.Bytes()
 }
 
+func (set Mysql56GTIDSet) Difference(other Mysql56GTIDSet) Mysql56GTIDSet {
+	if other == nil {
+		return set
+	}
+
+	// Make a fresh, empty set to hold the new value.
+	// This function is not supposed to modify the original set.
+	differenceSet := make(Mysql56GTIDSet)
+
+	for sid, intervals := range set {
+		otherIntervals, ok := other[sid]
+		if !ok {
+			// We didn't find SID in other set, so diff should include all intervals for sid unique to receiver.
+			differenceSet[sid] = intervals
+			continue
+		}
+
+		// Found server id match between sets, so now we need to subtract each interval.
+		var diffIntervals []interval
+		advance := func() bool {
```

Comment:
It might be possible to replace advance and advanceOther with a single func and passing the relevant []interval to it (and having it return the 0th element for later use).

Reply:
I had wanted to do that at first. The reason why I choose to use two different closures is that what we do for `advance` and `advanceOther` is different. In the case that we `advance` we always want to push a new interval onto `diffIntervals`. In the case that we `advanceOther` we do not. Unless I'm missing something, I don't think these can be merged into a single helper method.

Reply:
That is why I suggested returning the 0th element. As of now you always call `advance` with `intervals` and `advanceOther` with `otherIntervals`. Depending on which one you are operating on, you can do different things.
For instance, the first time you use a common advance func could look like this:
```
		if i, ok := advance(intervals); ok {
			diffIntervals = append(diffIntervals, i)
		}
		else {
			continue
		}
		if _, ok := advance(otherIntervals); !ok {
			differenceSet[sid] = intervals
			continue
		}
```
I'm not going to insist on this. If you feel the current version is more readable, that is fine.

Reply:
Ahhhh I see what you're saying. Yes, we could do it this way - although then we would have this kind of code in every switch branch. To me it reads cleaner to contain this logic into `advance` and `advanceOther` but this is certainly another way we could do it - and not worse. Likely just a matter of personal preference. Thanks for the feedback!